articles/external_qos_file.md

```diff
@@ -0,0 +1,402 @@
+The one in init options will be common to all nodes in the same contexts, and can be overridden by the one in node options.
+
+Optionally, we could also add in `rcl_node_options_t` a boolean to indicate to use the `rcl_qos_loaded_profiles_t` in init options if the profile was not found in the node specific loaded profiles object.
+
+### New ros argument
+
+A new argument will be needed to allow the user to specify which file to load.
+
+```bash
+ros2run <package_name> <exec_name> --ros-args --qos-file /path/to/qos/file
+```
+
+This loading mechanism will be equivalent to setting the `rcl_qos_loaded_profiles_t` member in init options.
+
+## Further extensions
+
+### Loading qos profiles in composable nodes
+
+The interface definition of the [LoadNode](https://github.com/ros2/rcl_interfaces/blob/master/composition_interfaces/srv/LoadNode.srv) service could be extended with a field to specify the desired `QoS` of the loaded node.
+
+In that way, the user loading the node is responsible of passing the desired QoS, instead of the node container.
+The message to be sent could be loaded from a qos profile file.
+
+#### Proposed interfaces
+
+```
+# EndpointQoS.msg
+
+string name  # This can be either a topic or service name
+uint8 type  # Endpoint type. An enum that can be a publisher/subscription/topic/service
+QoS qos  # Message that defines the qos
+```
+
+```
+# EndpointQoS.msg
+
+string name  # This can be either a topic or service name
+uint8 type  # Endpoint type. An enum that can be a publisher/subscription/topic/service
+QoS qos  # Message that defines the qos
+```
+
+```
+# composition_interfaces/srv/LoadNode.srv
+
+# Current request fields
+...
+# New field
+EndpointQoS[] requested_qos_profiles  # new field to allow loading user requested qos profiles
+---
+# Current response definition
+...
+```
+
+A `rcl_qos_loaded_profiles_t` can be constructed from the received message and passed to the node options of the node to be loaded.
+
+TBD: `QoS` message definition.
+
+### Support in launch files
+
+`launch_ros.actions.Node` and `launch_ros.actions.LoadComposableNode` actions could support a qos file argument in their constructors.
+
+## File format
+
+As showed in the above examples, both YAML or XML can correctly represent the desired file format.
+
+XML advantages:
+- Can be easily verified.
+YAML advantages:
+- It can be argued that the format is simpler and more readable than XML.
+- The format proposed in the above examples can easily be integrated with the current parameter file format.
+  That integration would lead in a unique "configuration file", that could be further extended in the future.
+  For example, passing remapping rules in it.
+
+## Explicitly allowing external configurability in entities
+
+Both node and init options could have a flag to ignore the new `--qos-file` argument.
+In a similar way, there could be an option for publishers/subscriptions/clients and services to allow side loading their qos or not.
```

Comment:
What does this mean? How does it relate to the QoS settings that the node receives?

Reply:
The node/context will have a structure of the loaded profiles.
When an entity is created and side loading qos is allowed, it will look up the loaded profiles in the node (or the global ones if the ones in the node aren't defined), and load the profile from there. That will only happen if the entity options allow side-loading too.

I will try to make that clearer in the document.

Reply:
The problem (for me) is that I'm still not clear what you mean by "side loading".

Reply:
With side loading, I mean loading the qos profile from a external file, or by another external way no specified in code (as the `rcl_qos_loaded_profiles_t` object could potentially be constructed in a different way).

I've rephrased in https://github.com/ros2/design/pull/296/commits/a0e635f3500e457bf56a36862651fe3bcf50b0a7, let me know if it's clearer now.

Reply:
Yep, that makes much more sense now. Thanks!